### ISSUE TITLE
Support 'quotedTemplate' in addition to 'StrLit' in block labels.

### ISSUE THREAD

**Issue by semtexzv** [hashicorp/hcl#600]
Currently, block label can be either identifier or string literal. I kind of want to support following pattern in a DSL I'm working on
```hcl
template {
    repeat = 10
    vars {
        prefix = "Hello"
    }
    generate {
        object "${vars.prefix}-${repeat}" {
            
        }
    }
}
```

But the HCL spec explicitly only allows strings or literals as object keys. 

Lifting the `quotedTemplate` from just subpattern of TemplateExpr into a rule that could be used as block label would greatly extend the usability, and reduce the need for patterns like `for_each` in terraform.

TL;DR: In [Syntax spec](https://github.com/hashicorp/hcl/blob/main/hclsyntax/spec.md#expressions):
Replace 
```ebnf 
Block        = Identifier (StringLit|Identifier)* "{" Newline Body "}" Newline;
```
with:
```ebnf
Block        = Identifier (StringLit|Identifier|QuotedTemplate)* "{" Newline Body "}" Newline;
```
